### PR TITLE
Function App specific checks for the network checker tool 

### PR DESCRIPTION
**Pull request by sidkri**
## Summary of changes
This PR introduces an additional network check workflow to the network checker tool that is available to all Function App resources.  The screen shot below shows a sample run of the tool against a Function App that is VNet integrated.

The PR also includes 3 files that can help with local development by obviating the need to install dependencies on a local machine other than the docker runtime itself.
* Dockerfile - builds a docker image that contains all dependencies needed to run the angular development environment
* buildimage.sh - run this once to build the image on your machine
* startcontainer.sh - run this to start the angular development environment
The README.md file is updated to include this optional step.

image

## Validation
I  validated this code by running it on a local angular server against a Function App that is VNet joined and tested via a combination of breaking the Function App (bad settings, firewalls etc.) and manipulating variables via the browser's debugging tool (e.g. add a breakpoint and change the value returned from an app setting to something that exercises a specific code path.  The latter approach was used to explore every single code path and validate that the rendered result matches what is expected with no exceptions in the rendered view or the browser's console.

### CODE REVIEW

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/connectionFailureFlow.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import {GetArmData, GetWebAppVnetInfo, GetSubnet, ResourcePermissionCheckManager} from './flowMisc.js';
+import {ResourcePermissionCheckManager, checkVnetIntegrationHealth, checkDnsSettingAsync, checkSubnetSizeAsync} from './flowMisc.js';
```
Comment (sidkri) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/connectionFailureFlow.js:1]: As the functionsFlow.js leverages a lot of the code from this file, I moved all relevant functions into flowMisc.js which is used by all flows.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/connectionFailureFlow.js [hunk 2]
```diff
@@ -41,18 +41,13 @@ export var connectionFailureFlow = {
...
-}
-
 async function runConnectivityCheck(hostname, port, dnsServers, diagProvider, lengthLimit) {
     var subChecks = [];
     var fellbackToPublicDns = false;
     var nameResolvePromise = (async function checkNameResolve() {
         var ip = null;
         var subChecks = [];
-        if (isIp(hostname)) {
+        if (diagProvider.isIp(hostname)) {
```
Comment (sidkri) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/connectionFailureFlow.js:50]: Avoiding duplicate code and the diagProvider.isIp() is more thorough

File: AngularApp/package.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     "applens": "node --max_old_space_size=8192 node_modules/@angular/cli/bin/ng serve applens --aot --optimization=false",
     "ncloud-applens-prod": "ng serve applens --configuration=nationalcloud-production --aot --optimization=false",
     "ncloud-applens-local": "ng serve applens --configuration=nationalcloud-local --aot --optimization=false",
-    "ssl": "node --max_old_space_size=8192 node_modules/@angular/cli/bin/ng serve app-service-diagnostics --aot --optimization=false --ssl --sslCert ./ssl/server.crt --sslKey ./ssl/server.key",
+    "ssl": "node --max_old_space_size=8192 node_modules/@angular/cli/bin/ng serve --host 0.0.0.0 app-service-diagnostics --aot --optimization=false --ssl --sslCert ./ssl/server.crt --sslKey ./ssl/server.key",
```
Comment (sidkri) [inline AngularApp/package.json:11]: This is needed when running within a docker container

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/configFailureFlow.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { GetArmData, GetWebAppVnetInfo, GetSubnet } from './flowMisc.js';
+import { getArmData, getWebAppVnetInfo, getSubnet } from './flowMisc.js';
```
Comment (sidkri) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/configFailureFlow.js:1]: Renamed to camelCase with lowercase first char to be consistent with the rest of the code base

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/flowMisc.js [hunk 5]
```diff
@@ -75,3 +74,1052 @@ export async function GetSubnet(armService, subnetResourceId) {
     }
     return subnetData;
 }
+
+export async function runKuduAccessibleCheck(diagProvider) {
```
Comment (sidkri) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/flowMisc.js:78]: All of these were moved from connectionFailureFlow.js to this file so they can be shared with functionsFlow.js

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-checks.component.ts [hunk 6]
```diff
@@ -97,6 +98,9 @@ export class NetworkCheckComponent implements OnInit, AfterViewInit {
             globals.messagesData.currentNetworkCheckFlow = null;
             var telemetryService = this._telemetryService;
             var networkCheckFlows = { connectionFailureFlow, configFailureFlow, learnMoreFlow }
+            if (this.siteInfo.kind.includes("functionapp")) {
```
Comment (cindylovescoding) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-checks.component.ts:101]: Currently for single tenant logic app, kind info will also include "functionapp", it will be be something like: "functionapp,workflowapp".
Could we update this condition to be if (!this.siteInfo.kind.includes("workflowapp") && this.siteInfo.kind.includes("functionapp"))?

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+        /**
+         * Function binding dependencies
+         **/
+        var checkFunctionBindingsPromise = (async () => {
+            // Explore the binding information of all functions in this function app to determine connection string
+            var functionsInfo = [];  // array of maps containing information about functions
+            var functionAppResourceId = siteInfo["resourceUri"];
+            var functionsList = await diagProvider.getArmResourceAsync(functionAppResourceId + "/functions");
+            functionsList.value.forEach(func => {
```
Comment (cindylovescoding) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:114]: Nitpick: can we do a null check for functionList.value != null before we call functionList.value.forEach()? Currently we are trying  hard to reduce any client side exceptions for Diagnose and Solve blade:)

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+                var functionInfo = { name: func.name, connectionStringProperties: [] };
+                func.properties.config.bindings.forEach(binding => {
```
Comment (cindylovescoding) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:116]: Same here to check if properties or config or bindings is null before referencing it:)

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+                    }
+                });
+                if(functionInfo.connectionStringProperties.length > 0) {
+                    functionsInfo.push(functionInfo);
+                }
+            });
+            var subChecksL1 = [];
+            var promises = functionsInfo.map(async (functionInfo) => {
+                var subChecksL2 = []; // These are the checks (and subchecks) for each binding of a function
+                var promises = functionInfo.connectionStringProperties.map(async (propertyName) => {
```
Comment (cindylovescoding) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:130]: Looks like we have a local var "promises" and a global var "promises" in the same block. It will not cause any issue because they are under different scope. But just for easier understanding, could we rename one of the "promises"  so when we do "await Promise.all(promises)" we know which one is called?

File: AngularApp/startcontainer.sh [hunk 10]
```diff
@@ -0,0 +1 @@
+docker run -it -v ${PWD}:/AngularApp -p 0.0.0.0:3000:3000 placeholder.repo:angularappimage bash
```
Comment (cindylovescoding) [inline AngularApp/startcontainer.sh:1]: Thank you Sid for adding this! It's very helpful!

**Review by cindylovescoding — APPROVED** [review]

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-checks.component.ts (hunk 6 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-checks.component.ts:101]: Good catch.  We should  test if the network checker works for logic app created function apps as well and expand support once we enable this tool for consumption and non-VNet joined function apps in the near future.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js (hunk 8 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:120]: Good point.  This also caused me to test and catch an edge case - when a Function App has no functions, the tool reported success in validating the bindings even though there were no functions.  Fixed both.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js (hunk 7 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:118]: good point.  Done.  I'll also look for additional paces that may need a null check.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js (hunk 9 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js]: Renamed them, should be easier to read now.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+        flowMgr.addView(checkFunctionBindingsPromise, "Checking all Function bindings...");
+
+        // General information about checks as positive will not always mean the app has no issues
+        flowMgr.addView(new InfoStepView({
+            infoType: 0,
+            title: "Explanation of the results and recommended next steps",
+            markdown: "Positive tests above indicate a network layer connection was successfully established between this app and the configured remote service."
+            + "\r\n\r\n" + "If the tests passed and your app is still having runtime connection failures with this endpoint, possible reasons could be:"
+            + "\r\n\r\n" + "-  There were authentication issues and the credentials involved have expired or are invalid. Only network connectivity was tested."
+            + "\r\n\r\n" + "-  The application setting was configured as a key vault reference and this diagnostics tool does not retrieve secrets from Key Vault.  Check application logs to debug further."
```
Comment (cachai2) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:175]: https://docs.microsoft.com/en-us/azure/app-service/app-service-key-vault-references

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+async function networkCheckKeyVaultReferenceAsync(propertyName, connectionString, dnsServers, diagProvider) {
+    var failureDetailsMarkdown = 'Please refer to <a href= "https://docs.microsoft.com/en-us/azure/app-service/app-service-key-vault-references#reference-syntax" target="_blank">this documentation</a> to configure the Key Vault reference correctly.'
+    var subChecks = [];
+    var hostPort = extractHostPortFromKeyVaultReference(connectionString);
+    if (hostPort.HostName != undefined && hostPort.Port != undefined) {
+        var connectivityCheckResult = await runConnectivityCheckAsync(hostPort.HostName, hostPort.Port, dnsServers, diagProvider, hostPort.HostName.length, failureDetailsMarkdown);
+        var maxCheckLevel = getMaxCheckLevel(connectivityCheckResult);
+        if(maxCheckLevel == 0) {
+            subChecks.push({
+                title: `Network access validation of connection strings configured as key vault references are currently not supported.  Network access to the Key Vault service referenced in the App Setting "${propertyName}" was verified. Recommend checking application logs for connectivity to the endpoint.`,
```
Comment (cachai2) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:226]: https://docs.microsoft.com/en-us/azure/app-service/app-service-key-vault-references

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+    }
+
+    // TCP Ping checks
+    var tcpPingResult = await tcpPingPromise;
+    var status = tcpPingResult.status;
+    if (status == 0) {
+        // Suppress successful checks to avoid clutter
+        //subChecks.push({ title: `TCP ping to ${hostname} was successful`, level: 0 });
+    } else if (status == 1) {
+        var markdown = `Connectivity test failed at TCP level for hostname **${hostname}** via resolved IP address ${resolvedIp}.  ` +
```
[the remark this answers is not in the document]
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:394]: Done, thanks!

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js (hunk 12 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:226]: already in the details markdown.  See "FailureDetailsMarkdown" which gets included in the subChecks

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+    var resultsMarkdown = nameResolveResult.checkResultsMarkdown;
+
+    var subChecks = [];
+
+    if (resolvedIp != hostname) {
+        hostname = hostname.length > lengthLimit ? hostname.substr(0, lengthLimit) + "..." : hostname;
+        if (resolvedIp == null) {
+            var markdown = "Results:"
+            resultsMarkdown.forEach(element => markdown += "\r\n- " + element);
+            markdown += `\r\n\r\nPossible reasons can be:` +
```
[the remark this answers is not in the document]
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:367]: Done

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+            var maxCheckLevel = getMaxCheckLevel(subChecksL1);
+            var title = maxCheckLevel == 0 ? "Network connectivity tests for all Function input/output bindings were successful." :
+                                             "Network connectivity tests failed for some Function input/output bindings.";
+            return new CheckStepView({ title: title, subChecks: subChecksL1, level: maxCheckLevel });
+        })();
+
+        flowMgr.addView(checkFunctionBindingsPromise, "Checking all Function bindings...");
+
+        // General information about checks as positive will not always mean the app has no issues
+        flowMgr.addView(new InfoStepView({
```
[the remark this answers is not in the document]
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:169]: yeah, I'd like to have a docs page that can be more dynamic than what is shown here but updating this tool is always relatively quick.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+            } else {
+                subChecksL1.push({
+                    title: `The app setting "${propertyName}" is not configured.  The Function App cannot work without this essential setting.`,
+                    level: 2,
+                    detailsMarkdown: failureDetailsMarkdown
+                });
+            }
+
+            // WEBSITE_CONTENTAZUREFILECONNECTIONSTRING
+            propertyName = "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING";
```
[the remark this answers is not in the document]
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:68]: WEBSITE_DNS_SERVER is already checked by "checkVnetIntegrationHealth()" in flowMisc.js:940
Let's discuss how to check the other two

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+
+    if (resolvedIp != hostname) {
+        hostname = hostname.length > lengthLimit ? hostname.substr(0, lengthLimit) + "..." : hostname;
+        if (resolvedIp == null) {
+            var markdown = "Results:"
+            resultsMarkdown.forEach(element => markdown += "\r\n- " + element);
+            markdown += `\r\n\r\nPossible reasons can be:` +
+            `\r\n\-  hostname **${hostname}** does not exist, please double check that the hostname is correct.` +
+            (dnsServers.filter(s => s != "").length == 0 ? "" : `\r\n\-  Your custom DNS server was used for resolving hostname, but there is no DNS entry on the server for **${hostname}**, please check your DNS server.`) +
+            "\r\n\-  If your target endpoint is an Azure service with Private Endpoint enabled, please check its Private Endpoint DNS Zone settings.";
```
[the remark this answers is not in the document]
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js]: Done.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js (hunk 11 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:175]: Am already warning the user of this when checking connection strings (and validating access to key vault itself).  Do you think we should link to the docs in this messaging too?

**Review by cindylovescoding — APPROVED** [review]

**Review by ShekharGupta1988 — APPROVED** [review]

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js [hunk 18]
```diff
@@ -0,0 +1,440 @@
...
+    }
+};
+
+function isKeyVaultReference(appSetting) {
+    return appSetting.includes("@Microsoft.KeyVault");
+}
+
+function getMaxCheckLevel(subChecks) {
+    var maxCheckLevel = 0;
+    subChecks.forEach(check => maxCheckLevel = Math.max(maxCheckLevel, check.level));
```
Comment (yifguoMSFT) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:205]: Does this work for level `info`?

**Review by yifguoMSFT — APPROVED** [review]

File: AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js (hunk 18 above)
Comment (sidkri) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js:205]: Not yet.  We ended up deciding to not change the icons at lower levels and keep them as either success, warning or failure.  At the top level check I convert success to info [here](https://github.com/Azure/Azure-AppServices-Diagnostics-Portal/blob/7ef31ae2dd071cb8d13a4c00df948ccb86a20f00/AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js#L112) and [here](https://github.com/Azure/Azure-AppServices-Diagnostics-Portal/blob/7ef31ae2dd071cb8d13a4c00df948ccb86a20f00/AngularApp/projects/app-service-diagnostics/src/app/shared/components/tools/network-checks/network-check-flows/functionsFlow.js#L173).
